File: app/src/main/java/com/bearcave/radio17/LoadArticle.java
package com.bearcave.radio17;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.AsyncTask;
import android.text.Html;
import android.text.Spanned;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.nostra13.universalimageloader.core.DisplayImageOptions;
import com.nostra13.universalimageloader.core.ImageLoader;
import com.nostra13.universalimageloader.core.ImageLoaderConfiguration;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.select.Elements;

import java.io.IOException;


/**
 * Created by miwas on 19.10.16.
 */

public class LoadArticle extends AsyncTask<String, Void, Elements> {

    ProgressDialog mProgressDialog;
    LinearLayout layout;
    Activity activity;
    int font_size;
    DisplayImageOptions options;
    ImageLoader imageLoader;
    static boolean clicked = false;
    static Intent intent;

    private IOException noInternetConnectionException;


    public LoadArticle(Activity activity, LinearLayout layout) {
        this.activity = activity;
        this.layout = layout;
        font_size = 17;

        options = new DisplayImageOptions.Builder()
                .showImageOnLoading(R.drawable.logo) // resource or drawable
                .cacheInMemory(true) // default => false
                .cacheOnDisk(true) // default => false
                .build();

        ImageLoaderConfiguration config = new ImageLoaderConfiguration.Builder(activity)
                .build();
        imageLoader = ImageLoader.getInstance();
        imageLoader.init(config);
    }


    private void enter(){
        TextView tv = new TextView(activity);
        tv.setText("\n");
        tv.setTextSize(font_size);
        layout.addView(tv);
    }

    private void addText(String text){
        TextView tv = new TextView(activity, null);
        tv.setText(text+"\n\n");
        tv.setTextSize( font_size );
        layout.addView(tv);
    }

    private void addText(Spanned text) {
        TextView tv = new TextView(activity, null);
        tv.setText(text);
        tv.setTextSize(font_size);
        layout.addView(tv);
    }

    @Override
    protected void onPreExecute() {

        super.onPreExecute();
        mProgressDialog = new ProgressDialog(activity);
        mProgressDialog.setMessage(activity.getString(R.string.loading));
        mProgressDialog.setIndeterminate(false);
        mProgressDialog.show();

    }

    @Override
    protected Elements doInBackground(String... urls) {

        Elements postContent = null;

        try {
            Document doc = Jsoup.connect(urls[0]).get();
            postContent = doc.getElementsByClass("post-content").first().children();
        } catch (IOException e) {
            noInternetConnectionException = e;
        }

        return postContent;
    }


    //this method checks what is in elements postContent and shows that on screen
    @Override
    protected void onPostExecute(Elements result) {

        if ( noInternetConnectionException != null){
            throw new NoInternetConnectionException();
        }

        boolean onlyText = true;
        mProgressDialog.dismiss();

        if (result == null) {
            return;
        }

        Elements elements;

        for (int j = 0; j < result.size(); j++) {

            elements = result.get(j).getAllElements();

            //removing unnecessery text
            for (int i = 0; i < elements.size(); i++) {
                String tmp = elements.get(i).tag().toString();

                if ( tmp == "img") {
                    onlyText = false;
                    i = elements.size();
                } else if ( tmp == "source"){
                    onlyText = false;
                    i = elements.size();
                    result.get(j).select("a").remove();
                } else if ( tmp =="style"){
                    result.get(j).remove();
                }
            }

            //displaying result on screen
            for (int i = 0; i < elements.size(); i++) {

                String tmp = elements.get(i).tag().toString();

                // IMAGE
                if (tmp == "img") {

                    if ( result.get(j).hasClass("gallery")) {
                        elements = result.get(j).getElementsByClass("gallery-item");

                        for ( int m = 0; m<elements.size(); m++){
                            ImageView ib = new ImageView(activity);
                            imageLoader.displayImage(elements.get(m).select("a").attr("href"), ib, options);
                            layout.addView(ib);
                            enter();
                        }
                    } else {
                        ImageView ib = new ImageView(activity);
                        imageLoader.displayImage(elements.get(i).absUrl("src"), ib, options);
                        layout.addView(ib);
                    }

                    if (result.get(j).hasText() ){
                        addText(result.get(j).text());
                    }

                    i = elements.size(); //  end this elements loop and go to the next result

                    //AUDIO
                } else if (tmp == "source") {

                    if (elements.get(i).attr("type").startsWith("audio")) {

                        final Button button = new Button(activity);
                        button.setText("PLAY");
                        final String pomUrl = elements.get(i).absUrl("src");
                        View.OnClickListener buttonListener = new View.OnClickListener() {

                            @Override
                            public void onClick(View v) {
                                if (clicked) {
                                    activity.stopService(intent);
                                    clicked = false;
                                } else {
                                    intent = new Intent(activity, PlayerService.class);
                                    intent.putExtra("audio_src", pomUrl);
                                    clicked = true;
                                    activity.startService(intent);
                                }
                            }
                        };
                        layout.addView(button);
                        button.setOnClickListener(buttonListener);

                        if (result.get(j).hasText() ){
                            addText(result.get(j).text());
                        }

                        i = elements.size(); // go to next result's element

                    }
                } else if ( tmp == "style"){}
                //TEXT
                else if ( onlyText ){
                    addText(Html.fromHtml(result.get(j).toString()));
                    i = elements.size();

                }
            }
            onlyText = true;
        }
    }
}

File: app/src/main/java/com/bearcave/radio17/articles/LoadArticle.kt
package com.bearcave.radio17.articles

import android.app.Activity
import android.app.ProgressDialog
import android.content.Context
import android.os.AsyncTask
import android.text.Html
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView

import com.bearcave.radio17.exceptions.NoInternetConnectionException
import com.bearcave.radio17.R
import com.nostra13.universalimageloader.core.DisplayImageOptions
import com.nostra13.universalimageloader.core.ImageLoader
import com.nostra13.universalimageloader.core.ImageLoaderConfiguration
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import java.io.IOException
import java.util.*


/**
 * Loads article from url given from parent activity and displays it on screen.
 * @author Michał Wąsowicz
 */
class LoadArticle(internal val context: Context, internal val layout: LinearLayout) : AsyncTask<String, Void, Element>() {

    internal val mProgressDialog: ProgressDialog = ProgressDialog(context)
    internal val font_size: Int = 17

    internal val options: DisplayImageOptions = DisplayImageOptions.Builder()
            .showImageOnLoading(R.drawable.logo)
            .cacheInMemory(true)
            .cacheOnDisk(true)
            .build()
    internal val imageLoader: ImageLoader = ImageLoader.getInstance()
    internal val tagToViewMap = HashMap<String, (Element)->View>()

    private var noInternetConnectionException: IOException? = null


    init {
        val config = ImageLoaderConfiguration.Builder(context)
                .build()
        imageLoader.init(config)

        tagToViewMap.put("p", this::addText)
        tagToViewMap.put("strong", this::addText)
        tagToViewMap.put("b", this::addText)
        tagToViewMap.put("span", this::addText)
        tagToViewMap.put("i", this::addText)
        tagToViewMap.put("em", this::addText)
        tagToViewMap.put("div", this::addText)

        tagToViewMap.put("img", this::addImage)
    }

    override fun onPreExecute() {
        super.onPreExecute()
        mProgressDialog.setMessage(context.getString(R.string.loading))
        mProgressDialog.isIndeterminate = false
        mProgressDialog.show()
    }

    override fun doInBackground(vararg urls: String): Element? {
        try {
            val doc = Jsoup.connect(urls[0]).get()
            return doc.getElementsByClass("post-content").first()
        } catch (e: IOException) {
            noInternetConnectionException = e
        }

        return null
    }

    /**
     * Checks what is in element postContent and shows that on screen
     * @exception NoInternetConnectionException
     */
    override fun onPostExecute(result: Element) {
        mProgressDialog.dismiss()
        if (noInternetConnectionException != null) {
            throw NoInternetConnectionException()
        }

        extractElement(result)
    }

    // TODO: podzielic layout.addView(getView(element)) na dwie czesci: przed wywolaniem rekurencyjnym i po.
    private fun extractElement(element: Element){
        val elements = element.children()
        addView(element)
        for ( child in elements){
            extractElement(child)
        }
    }

    private fun addView(elementView: Element){
        val view = getView(elementView)
        if(view != null) {
            layout.addView(view)
        }
    }

    private fun getView(element: Element): View? {
        return tagToViewMap[element.tagName()]?.invoke(element)
    }

    private fun addText(element: Element): View {
        val tv = TextView(context)
        tv.text = Html.fromHtml("<${element.tagName()}>${element.ownText()}</${element.tagName()}>")
        tv.textSize = font_size.toFloat()
        return tv
    }

    private fun addImage(element: Element): View{
        val ib = ImageView(context)
        imageLoader.displayImage(element.absUrl("src"), ib, options)
        return ib
    }

    private fun addPlayer(element: Element): View{
        val ib = ImageView(context)
        imageLoader.displayImage(element.absUrl("src"), ib, options)
        return ib
    }
}